File: Snake/Snake/Logic/Cell.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Snake.Logic
{
    public class Cell
    {
        public int X { get; set; }
        public int Y { get; set; }
        public List<Tuple<Position, Direction>> Move { get; set; }

        public Direction Heading { get; set; }

        public Cell(int x, int y,Direction d)
        {
            X = x;
            Y = y;
            Heading = d;
            Move = new List<Tuple<Position, Direction>>();
        }

        private int[] xs = {-1, 0, 1, 0};
        private int[] ys = {0, 1, 0, -1};

        public bool Moving(Board board, List<Cell> body)
        {
            if (!Moved(Heading, board.Rows, board.Columns,body))
                return false;

            if(Move.Count > 0)
            {
                if (Move[0].Item1.X == X && Move[0].Item1.Y == Y)
                {
                    Heading = Move[0].Item2;
                    Move.RemoveAt(0);
                }
            }

            return true;

        }

        public bool Moved(Direction d, int rows, int columns, List<Cell> body)
        {
            int newX = X + xs[(int) d];
            int newY = Y + ys[(int) d];

            if (body.Any(x => x.X == newX && x.Y == newY))
                return false;

            if (!Tools.In(newX, newY, rows, columns))
                return false;

            X = newX;
            Y = newY;

            return true;

        }

        public Cell GetDirection(Direction heading,int rows,int columns)
        {
            List<Direction> posible = new List<Direction>(){};

            posible.Add(DirectionTools.Front(heading));

            DirectionTools.Sides(heading).ForEach(posible.Add);
            
            foreach (var d in posible)
            {
                int newX = X + xs[(int)d];
                int newY = Y + ys[(int)d];

                if (Tools.In(newX, newY, rows, columns))
                {
                    Cell c = new Cell(newX,newY,heading);

                    //c.Move.Add(new Tuple<Position, Direction>(new Position(X, Y), heading));

                    foreach (var m in Move)
                    {
                        c.Move.Add(new Tuple<Position, Direction>(new Position(m.Item1.X, m.Item1.Y), m.Item2));
                    }

                    //c.Move.Add(new Tuple<Position, Direction>(new Position(X, Y), heading));

                    return c;
                }
            }

            return null;


        }

        
    }
}

File: Tools/Cell.js

const xs = [-1, 0, 1, 0];
const ys = [0, 1, 0, -1];

function Cell(x, y,d)
{
    this.X = x;
    this.Y = y;
    this.Heading = d;
    this.Move = [];

    
}

Cell.prototype.X = function()
{
    return this.X;
}

Cell.prototype.Y = function()
{
    return this.Y;
}

Cell.prototype.Move = function()
{ 
    return this.Move; 
}

Cell.prototype.Heading = function()
{ 
    return this.Heading 
}

Cell.prototype.Moving = function (board, body)
{
    if (!this.Moved(this.Heading, board.Rows(), board.Columns(),body))
        return false;

    if(this.Move.length > 0)
    {
        if (this.Move[0].Item1().X == this.X && this.Move[0].Item1().Y == this.Y)
        {
            this.Heading = this.Move[0].Item2();
            this.Move.splice(0,1);
        }
    }

    return true;

}

Cell.prototype.Moved = function(d, rows, columns, body)
{
    var newX = this.X + xs[DirectionToInt[d]];
    var newY = this.Y + ys[DirectionToInt[d]];

    if (this.Any(body,newX,newY))
        return false;

    var t = new Toolsc();

    if (!t.In(newX, newY, rows, columns))
        return false;

    this.X = newX;
    this.Y = newY;

    return true;

}

Cell.prototype.Any = function(body,newX,newY)
{
    for(var c in body)
    {
        var cell = body[c];
        
        if(cell.X == newX && cell.Y == newY)
            return true;
    }

    return false;
}

Cell.prototype.GetDirection = function(heading, rows, columns)
{
    var posible = [];

    var di = new DirectionTools();

    posible.push(di.Front(heading));

    var si = di.Sides(heading); 

    for(var s in si)
        posible.push(si[s]);    

    var t = new Toolsc(); 
    
    for (d in posible)
    {
        var dir = posible[d];

        var newX = this.X + xs[DirectionToInt[dir]];
        var newY = this.Y + ys[DirectionToInt[dir]];        

        if (t.In(newX, newY, rows, columns))
        {
            var c = new Cell(newX,newY,heading);

            //c.Move.Add(new Tuple<Position, Direction>(new Position(X, Y), heading));

            for (m in this.Move)
            {
                var mo = this.Move[m];

                c.Move.push(new Pair(new Position(mo.Item1().X, mo.Item1().Y), mo.Item2()));
            }

            //c.Move.Add(new Tuple<Position, Direction>(new Position(X, Y), heading));

            return c;
        }
    }

    return null;

}